roachpb/data.go

```diff
@@ -797,20 +809,27 @@ func (t Transaction) Short() string {
 	return t.ID.String()[:8]
 }
 
-// Add adds the given NodeID to the interface (unless already present)
-// and restores ordering.
-func (s *NodeList) Add(nodeID NodeID) {
-	if !s.Contains(nodeID) {
-		s.Nodes = append(s.Nodes, nodeID)
-		sort.Sort(NodeIDSlice(s.Nodes))
+// UpdateUncertainty is to remember a timestamp off a node's clock for future
+// operations on that node. When multiple calls are made for a single nodeID,
+// the lowest timestamp prevails.
+func (t *Transaction) UpdateUncertainty(nodeID NodeID, maxTS Timestamp) {
+	if t.MaxTimestamps == nil {
+		t.MaxTimestamps = make(map[NodeID]Timestamp)
+	}
+	if ts, ok := t.MaxTimestamps[nodeID]; !ok || maxTS.Less(ts) {
+		t.MaxTimestamps[nodeID] = maxTS
 	}
 }
 
-// Contains returns true if the underlying slice contains the given NodeID.
-func (s NodeList) Contains(nodeID NodeID) bool {
-	ns := s.Nodes
-	i := sort.Search(len(ns), func(i int) bool { return NodeID(ns[i]) >= nodeID })
-	return i < len(ns) && NodeID(ns[i]) == nodeID
+// GetUncertainty returns the lowest timestamp recorded for the given node.
```

Comment:
I do think recasting all of these function names (and comments) as dealing with max timestamps will be more straightforward.